test/vega_lite/convert_test.exs

```diff
@@ -1,3 +1,223 @@
 defmodule VegaLiteConvertTest do
   use ExUnit.Case
+
+  alias VegaLite, as: Vl
+  alias VegaLite.Convert
+
+  @data [
+    %{"height" => 170, "weight" => 80, "width" => 10, "unused" => "a"},
+    %{"height" => 190, "weight" => 85, "width" => 20, "unused" => "b"}
+  ]
+
+  describe "save!/3" do
+    @tag :tmp_dir
+    test "should write a file to filesystem for all of the supported types", %{tmp_dir: tmp_dir} do
+      supported_file_types = [
+        :json,
+        :html,
+        :png,
+        :svg,
+        :pdf,
+        :jpeg,
+        :jpg
+      ]
+
+      vl = generate_vl()
+
+      Enum.each(supported_file_types, fn file_type ->
+        path = Path.join(tmp_dir, "test_file.#{file_type}")
+        Convert.save!(vl, path)
+        assert File.exists?(path)
+      end)
+    end
+
+    @tag :tmp_dir
+    test "should raise an error when an invalid file name is provided", %{tmp_dir: tmp_dir} do
+      vl = generate_vl()
+      path = Path.join(tmp_dir, "test_file.invalid")
+
+      assert_raise ArgumentError,
+                   "unsupported export format, expected :json, :html, :png, :svg, :pdf, :jpeg or :jpg got: :invalid",
+                   fn ->
+                     Convert.save!(vl, path)
+                   end
+    end
+
+    @tag :tmp_dir
+    test "should generate a file with a nonstandard extension if explicitly provided", %{
+      tmp_dir: tmp_dir
+    } do
+      vl = generate_vl()
+
+      path = Path.join(tmp_dir, "test_file.pngv2")
+      Convert.save!(vl, path, format: :png)
+      assert File.exists?(path)
+    end
+  end
+
+  describe "to_json/2" do
+    test "should return the spec as VegaLite formatted JSON" do
+      vl = generate_vl()
+
+      assert Convert.to_json(vl) =~ "https://vega.github.io/schema/vega-lite/v5.json"
+    end
+
+    test "should return the spec as Vega formatted JSON" do
+      vl = generate_vl()
+
+      assert Convert.to_json(vl, target: :vega) =~ "https://vega.github.io/schema/vega/v5.json"
+    end
+  end
+
+  describe "to_svg/1" do
+    test "should return the spec as VegaLite formatted JSON" do
+      vl = generate_vl()
+      svg = Convert.to_svg(vl)
+
+      assert svg =~ "xmlns=\"http://www.w3.org/2000/svg\""
+    end
+  end
+
+  describe "to_pdf/1" do
+    test "should generate a PDF document given a VegaLite spec" do
+      vl = generate_vl()
+
+      pdf =
+        vl
+        |> Convert.to_pdf()
+        |> :binary.list_to_bin()
```

Comment:
The return value should already be a binary, no?

Reply:
For a few of the functions vl-convert-rs returns `Result<Vec<u8>, AnyError>` as opposed to `Result<String, AnyError>` and Rustler automatically serializes that as a list (https://docs.rs/vl-convert-rs/1.7.0/vl_convert_rs/converter/struct.VlConverter.html#method.vegalite_to_pdf)

Reply:
Got it, in that case let's change the function typespecs to say they return `iodata()`.

We could convert to a binary in Rust upfront, but I'm not sure if we can do it zero-copy, in which case I think `iodata()` makes more sense. cc @josevalim

Reply:
Wait, if we return `Vec<u8>` and Rustler encodes that as a list, that's obviously also copying, and it's actually worse. So we should always return a binary, yeah.

Reply:
I tried playing around with this for a little bit but I am not finding a good way to coerce the types from `Vec<u8> -> String`. Then again I am very new to Rust 😆.

Reply:
I tried doing `String::from_utf8(png).unwrap()` as was suggested in a few forum posts but it looks like that is returning `Utf8Error` and the unwrap panics.

Reply:
When rustler converts `String` to term it actually copies too. So we want to convert from vec to binary directly. Try this:
Try this:

```rust
fn encode_vec_result(env: Env, result: Result<Vec<u8>, Error>) -> Result<Binary, String> {
    return match result {
        Ok(vec) => {
            let mut bin = OwnedBinary::new(vec.len()).expect("allocation failed");
            bin.as_mut_slice().copy_from_slice(vec.as_slice());
            Ok(Binary::from_owned(bin, env))
        }
        Err(err) => Err(err.to_string()),
    };
}
```

and add `env: Env` as first argument to the NIFs.

On a sidenote, I would also reuse the logic for the string cases:

```rust
fn encode_string_resultresult: Result<String, Error>) -> Result<String, String> {
    return match result {
        Ok(string) => Ok(string),
        Err(err) => Err(err.to_string()),
    };
}
```

Reply:
So then you call it either as `return encode_string_result(svg_result);` or `return encode_vec_result(env, png_result);`.

Reply:
I understand now. Thanks for the in depth explanation! Commit incoming with these changes.

Reply:
Commit 8c14c89509c764a554a6202dcd9d6f2f900a01eb should address this comment.